File: backend/app/schemas/crucible_charge.py
"""
Schemas del documento de crisol (#107 D2) — el 4º "ítem" de Salidas de Plomo.

`charge` = traslado a crisoles (kg de crudo pasan de la etapa horno a la etapa
crisol de la cuenta intersede); `dross_return` = retorno de dross al horno (kg
vuelven de crisol a horno y se causa una nueva maquila). `discharge` — el cierre
de refinacion de la spec — NO se acepta: el Literal lo corta en 422 (en el modelo
de Hugo el crisol no se cierra, baja al vender puro y al devolver dross).

`quantity_kg` son kg de PLOMO (lo que cambia de etapa), no kg fisicos del
material: el documento es control de donde esta el plomo (Hugo 28-ago), no
mueve inventario.
"""
from datetime import datetime
from decimal import Decimal
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.utils.dates import BusinessDate

CrucibleEventType = Literal["charge", "dross_return"]


class CrucibleChargeCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    event_type: CrucibleEventType
    warehouse_id: UUID
    material_id: UUID
    quantity_kg: Decimal = Field(..., gt=0, description="kg de PLOMO que cambian de etapa")
    date: BusinessDate
    notes: Optional[str] = Field(None, max_length=1000)


class CrucibleChargeAnnul(BaseModel):
    reason: str = Field(..., min_length=3, max_length=500)


class CrucibleChargeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    charge_number: int
    label: str
    event_type: str
    warehouse_id: UUID
    warehouse_name: Optional[str] = None
    material_id: UUID
    material_code: Optional[str] = None
    material_name: Optional[str] = None
    quantity_kg: Decimal
    date: datetime
    notes: Optional[str] = None
    status: str
    maquila_amount: Decimal = Decimal("0")
    annulled_reason: Optional[str] = None
    annulled_at: Optional[datetime] = None
    annulled_by_name: Optional[str] = None
    created_by_name: Optional[str] = None
    created_at: datetime
    warnings: list[str] = Field(default_factory=list)


class CrucibleChargeListResponse(BaseModel):
    items: list[CrucibleChargeResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/types/crucible-charge.ts
// Documentos de crisol (#107 D2): el 4º "ítem" de Salidas de Plomo.
//
// `charge` = traslado a crisoles (horno −kg / crisol +kg, sin pesos);
// `dross_return` = retorno de dross al horno (crisol −kg / horno +kg, y se causa
// la maquila del reproceso). Ninguno mueve inventario: es control de DÓNDE está
// el plomo dentro de la deuda de planta con Circunvalar.
//
// ⚠️ Los Decimal del backend llegan como STRING (bloqueante (b) de #93): los
// montos se declaran `string | number` y se leen con `num()`.

import { num } from "./willard-delivery";

export { num };

export type CrucibleEventType = "charge" | "dross_return";
export type CrucibleChargeStatus = "confirmed" | "annulled";

export const CRUCIBLE_EVENT_LABELS: Record<CrucibleEventType, string> = {
  charge: "Traslado a crisoles",
  dross_return: "Retorno de dross al horno",
};

export const CRUCIBLE_EVENT_COLORS: Record<CrucibleEventType, string> = {
  charge: "bg-orange-100 text-orange-800",
  dross_return: "bg-stone-200 text-stone-800",
};

export const CRUCIBLE_STATUS_LABELS: Record<CrucibleChargeStatus, string> = {
  confirmed: "Registrado",
  annulled: "Anulado",
};

export interface CrucibleCharge {
  id: string;
  charge_number: number;
  /** "Crisol #n" — usar este, no charge_number. */
  label: string;
  event_type: CrucibleEventType;
  warehouse_id: string;
  warehouse_name: string | null;
  material_id: string;
  material_code: string | null;
  material_name: string | null;
  quantity_kg: string | number;
  date: string;
  notes: string | null;
  status: CrucibleChargeStatus;
  /** Maquila del reproceso causada por un retorno de dross ($0 en un traslado). */
  maquila_amount: string | number;
  annulled_reason: string | null;
  annulled_at: string | null;
  annulled_by_name: string | null;
  created_by_name: string | null;
  created_at: string;
  /** Advertencias no bloqueantes (#17/#76) — se leen de la RESPUESTA (#100 D4d). */
  warnings?: string[];
}

export interface CrucibleChargeListResponse {
  items: CrucibleCharge[];
  total: number;
  page: number;
  page_size: number;
}

export interface CrucibleChargeCreate {
  event_type: CrucibleEventType;
  warehouse_id: string;
  material_id: string;
  quantity_kg: string;
  date: string;
  notes?: string | null;
}
